### ISSUE TITLE
Delta Bundles on large Datasets/Snapshot-Bundles - Memory usage

### ISSUE THREAD

**Issue by smilowski** [open-policy-agent/opa#6079]
## Short description 

Hi All,
We seem to experience quite strange memory spikes on OPA loading **delta** bundles with larger datasets in InMemory storage.

Let us assume we have the following Structure in OPA ( loaded previously with a snapshot bundle ): 
`{
  "list_of_persons": {
    "person1": {
      "relatedPerson1": {
        "relationType": [
          [
            "validFrom",
            "validTo"
          ]
        ]
      }
    }
  }`
The dataset is quite large (OPA runtime -> 1.8 Gb)

Once patches/delta bundles of the following form hit OPA, there is a x2, x3 memory spike.
`
{
  "data": [
    {
      "op": "upsert",
      "path": "/list_of_persons/nonExistingPerson1/relatedPerson1",
      "value": {
        "relationType": [
           [
            "validFrom",
            "validTo"
          ]
        ]
      }
    }
}
`

I hope I am not mistaken in my analysis but, through debugging locally I was able to see that through an 'upsert' operation on a non-existing 'leaf' there is a DeepCopy of the whole InMem Storage 

https://github.com/open-policy-agent/opa/blob/main/storage/inmem/txn.go#L210

OPA calls the Read method with _"/list_of_persons/nonExistingPerson1"_, which returns NotFoundErr. OPA tries to create _"/list_of_persons/nonExistingPerson1"_ but needs to firstly check that _"/list_of_persons"_ exists. The **Read** method is called with: path = 'list_of_persons', which in turn makes a DeepCopy of 'list_of_persons'

Some additional info:
opa_version: latest (master checked-out)

## Expected behavior
My expectation would be that OPA reconciles/merges only the updates in the transaction/patch itself and not the whole InMem storage.
In other words, that it does not produce a DeepCopy of the whole internal storage on every non-existing leaf.

Hope I could make my problem clear and would appreciate some feedback.
Have a great day

**Comment by ashutosh-narkar** [comment]
The logic behind making a deep copy of the data during a read seems to handle the merging of [updates](https://sourcegraph.com/github.com/open-policy-agent/opa@f0b351d683da3c856f20c5181f7824c2e46f8bd7/-/blob/storage/inmem/txn.go) and not to modify the original.  We'll need to try this out but in your example were the any updates in the list [here](https://sourcegraph.com/github.com/open-policy-agent/opa@f0b351d683da3c856f20c5181f7824c2e46f8bd7/-/blob/storage/inmem/txn.go?L184)? Also if you have a sample snapshot and delta bundles handy and could share those, that would be hepful.

**Comment by smilowski** [comment]
hi @ashutosh-narkar 
Exacly, the delta contains mutliple updates to the same branch, so this loop is triggered and merges are applied to the transaction. 

I've generated example snapshot and delta bundles, which show this behavior. 

Snapshot contains 1Mio. relations ( OPA Runtime ~1Gb after activation )
[snapshot.tar.gz](https://github.com/open-policy-agent/opa/files/11980231/snapshot.tar.gz)

Delta contains only 20 'upsert' operations, where some branches are new.
[delta.tar.gz](https://github.com/open-policy-agent/opa/files/11980230/delta.tar.gz)

I've not tried Disk Storage, we are purely on in-memory storage

**Comment by smilowski** [comment]
hi @ashutosh-narkar 
Was wondering if you had some time to investigate this issue.

Seems like a tricky situation, due to the algorithm design. 
The algorithm seems to copy the internal storage and update it inside of the transaction scope. Hence the requirement for the [DeepCopy](https://github.com/open-policy-agent/opa/blob/main/storage/inmem/txn.go#L210) operation. 
It then applies the new structures via write operation once all patches are processed. I assume, this design makes an all-or-nothing update easier to implement.

Was wondering if an introduction of an 'Exists' method on the Store interface would alleviate the problem.
It seems the result of Read operation at those 2 locations is obsolete. 
- https://sourcegraph.com/github.com/open-policy-agent/opa@f0b351d683da3c856f20c5181f7824c2e46f8bd7/-/blob/bundle/store.go?L934
- https://sourcegraph.com/github.com/open-policy-agent/opa@f0b351d683da3c856f20c5181f7824c2e46f8bd7/-/blob/storage/storage.go?L65

The [Read](https://sourcegraph.com/github.com/open-policy-agent/opa@f0b351d683da3c856f20c5181f7824c2e46f8bd7/-/blob/storage/inmem/txn.go?L176) method is however responsible for the reconciliation of the updates so the new Exists method would need to somehow do the same job

Do you have any thoughts?

**Comment by ashutosh-narkar** [comment]
Hey @smilowski, I haven't had a chance to look into this. From looking at the code, not sure an `Exists` method would have any benefit here. As you point out those Read calls at the 2 places could be improved and that may help with performance. In terms of improvements to the Read method itself, that would be a more involved change and need further investigation. 

Also would be worth trying out disk storage to see if any performance benefit from that.
